File: String/20ValidParentheses.cpp
/**
 Solution:
 Use stack.
*/
class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        for(int i=0; i<s.size(); i++){
            if(s[i]=='{' || s[i]=='(' || s[i]=='[') stack.push_back(s[i]);
            if(s[i]=='}' || s[i]==')' || s[i]==']'){
                if(stack.empty()) return false;
                char temp = stack.back();
                if((s[i]=='}' && temp=='{') || (s[i]==')' && temp=='(') || (s[i]==']' && temp=='[')) stack.pop_back();
                else return false;
            }
        }
        return stack.empty() ? true : false;
    }
};

File: String/20validParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        arr = []
        
        for i in s:
            if i == '(' or i == '{' or i == '[':
                arr.append(i)
            if (i == ')' or i == '}' or i == ']') and not arr:
                print(1)
                return False
            if i == ')':
                if arr.pop() != '(':
                    print(2)
                    return False
            elif i == ']':
                if arr.pop() != '[':
                    print(3)
                    return False
            elif i == '}':
                if arr.pop() != '{':
                    print(4)
                    return False
            else:
                pass
        return arr == []